src/cracker.c

```diff
@@ -818,6 +821,19 @@ static int crk_password_loop(struct db_salt *salt)
 	if (fp_fix_state)
 		fp_fix_state();
 
+	if (kpc_warn && crk_key_index < crk_params->min_keys_per_crypt) {
+		static int last_warn_kpc;
+
+		if (!mask_increments_len && last_warn_kpc != crk_key_index) {
+			last_warn_kpc = crk_key_index;
+			fprintf(stderr,"Warning: %u: Only %d candidates left, minimum %d "
+			        "needed for performance.\n%s",
+			        NODE, crk_key_index, crk_params->min_keys_per_crypt,
```

Comment:
Don't we normally print the node number before any messages (so "%u: Warning: ...") and only when there's more than one node?

Reply:
Yes, I actually have a pending commit that adds a `log_print(int verbosity, bool main_only, char *format, ...)` function with all that functionality built-in and that's why it's not here. That change is much harder to get right than I could ever anticipate. This PR won't be merged right now anyway because I've seen some caveats that I need to dig into.